cmd/skopeo/sync.go

```diff
@@ -0,0 +1,509 @@
+	if err != nil {
+		return nil, fmt.Errorf("Invalid image name %s: %v", imgName, err)
+	}
+	valid, err := isValidTransport(ref.Transport())
+	if !valid {
+		return nil, fmt.Errorf("Invalid transport")
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	return ref, nil
+}
+
+// Builds the final destination of the image:
+// eg: given destination `docker://my-registry.local.lan` and src `docker://docker.io/busybox:stable`
+// the final destination is going to be docker://my-registry.local.lan/docker.io/busybox:
+func buildFinalDestination(rootSrc, srcRef types.ImageReference, globalDest string) (types.ImageReference, error) {
+	var dest string
+
+	if srcRef.Transport() == transports.Get("docker") {
+		dest = fmt.Sprintf("%s/%s", globalDest, srcRef.DockerReference())
+	} else {
+		// It's `dir` transport. In that case`globalDest` uses the `docker://` transport because
+		// we don't allow `dir` -> `dir` sync operations.
+		relPath := strings.TrimPrefix(srcRef.StringWithinTransport(), rootSrc.StringWithinTransport())
+		// we cannot use `filepath.Join` against a "vanilla" globalDest because it will change
+		// `docker://` into `docker:/` breaking the reference validations later on
+		dest = fmt.Sprintf(
+			"docker://%s",
+			filepath.Join(strings.TrimPrefix(globalDest, "docker://"), relPath))
+	}
+
+	if strings.HasPrefix(dest, "dir:") {
+		// the final directory holding the image must exist otherwise
+		// the directory ImageReference instance won't be created
+		tgtDir := filepath.Dir(strings.TrimPrefix(dest, "dir:"))
+
+		if _, err := os.Stat(tgtDir); err != nil {
+			if os.IsNotExist(err) {
+				if err := os.MkdirAll(tgtDir, 0755); err != nil {
+					return nil, fmt.Errorf("Error creating directory for image %s: %v",
+						tgtDir,
+						err)
+				}
+			} else {
+				return nil, fmt.Errorf("Error while checking existance of directory %s: %v",
+					tgtDir,
+					err)
+			}
+		}
+	}
+
+	return getImageReference(dest)
+}
+
+func getImageTags(ctx context.Context, sysCtx *types.SystemContext, imgRef types.ImageReference) (tags []string, retErr error) {
+	img, err := imgRef.NewImage(ctx, sysCtx)
+	if err != nil {
+		return tags, err
+	}
+	defer func() {
+		if err := img.Close(); err != nil {
+			retErr = errors.Wrapf(retErr, fmt.Sprintf("(could not close image: %v) ", err))
+		}
+	}()
+	if dockerImg, ok := img.(*docker.Image); ok {
+		logrus.WithFields(logrus.Fields{
+			"image": dockerImg.SourceRefFullName(),
+		}).Info("Getting tags")
+		tags, retErr = dockerImg.GetRepositoryTags(context.Background())
+		if retErr != nil {
+			// some registries may decide to block the "list all tags" endpoint
+			// gracefully allow the inspect to continue in this case. Currently
+			// the IBM Bluemix container registry has this restriction.
+			if !strings.Contains(retErr.Error(), "401") {
+				return tags, fmt.Errorf("Error determining repository tags: %v", retErr)
+			}
+			logrus.Warn("Registry disallows tag list retrieval; skipping")
+		}
+	}
```

Comment:
Maybe I need more tea, but should you return 'tags, nil' here and/or at line 152?

Reply:
Oh shoot, forgot golang deals with bare returns.  This code is OK as is, but I'd prefer specific values being returned.  I find it more readable.  My $.02, your mileage may vary.